packages/mapviewer/src/modules/menu/components/advancedTools/ImportCatalogue/ProviderList.vue

```diff
@@ -1,49 +1,221 @@
+            prefix = prefix.slice(0, -1)
+            if (!prefix) break
+        }
+    }
+    // Ensure the prefix ends with a '/'
+    return prefix.endsWith('/') ? prefix : prefix.slice(0, prefix.lastIndexOf('/') + 1)
+}
+
+// Function to build the tree node structure
+// This function takes a base URL and an array of providers and returns a tree node
+// structure. If there is only one provider, it returns a URL node. If there are multiple
+// providers, it groups them by their common prefix and creates a group node with
+// sub-group nodes for each unique prefix.
+//  It also set the name as the text to be shown in the UI
+function buildTreeNode(baseUrl, providers) {
+    if (providers.length === 1) {
+        return {
+            id: baseUrl,
+            name: providers[0].htmlDisplay,
+            type: 'url',
+            url: providers[0].url,
+            emphasize: providers[0].emphasize,
+        }
+    }
+
+    const urls = providers.map((provider) => provider.url)
+    const commonPrefix = getLongestCommonPrefix(urls)
+    const subGroups = {}
+
+    providers.forEach((provider) => {
+        const relativeUrl = provider.htmlDisplay.replace(commonPrefix || baseUrl, '')
+        const subGroupKey = relativeUrl.split('/')[0]
+
+        if (!subGroups[subGroupKey]) {
+            subGroups[subGroupKey] = []
+        }
+        subGroups[subGroupKey].push({
+            ...provider,
+            relativeUrl: relativeUrl.replace(subGroupKey + '/', ''),
+        })
+    })
+
+    return {
+        id: baseUrl,
+        name: commonPrefix || baseUrl,
+        type: 'group',
+        expanded: filterApplied,
+        children: Object.entries(subGroups).map(([key, subGroupProviders]) => {
+            if (subGroupProviders.length === 1) {
+                return {
+                    id: subGroupProviders[0].url,
+                    name: key + '/' + subGroupProviders[0].relativeUrl,
+                    type: 'url',
+                    url: subGroupProviders[0].url,
+                    emphasize: subGroupProviders[0].emphasize,
+                }
+            }
+            return {
+                id: `${baseUrl}-${key}`,
+                name: key,
+                type: 'group',
+                expanded: false,
+                children: subGroupProviders.map((provider) => ({
+                    id: provider.url,
+                    name: provider.relativeUrl,
+                    type: 'url',
+                    url: provider.url,
+                    emphasize: provider.emphasize,
+                })),
+            }
+        }),
+    }
+}
+
+// Set tabindex for each node in the tree with DFS strategy so that the tab
+// index is in correct order visually
+function addTabIndex(treeData) {
+    let index = 0
+
+    function dfs(node) {
+        node.tabindex = index++
+        if (node.children && node.children.length > 0) {
+            node.children.forEach((child) => dfs(child))
+        }
+    }
```

Comment:
why nesting this function? You can maybe give the `index` as a param to the function instead, and return the resulting index instead of mutating it inside the function

Reply:
Would also make sense to add a little JSDoc that explains what this function does, `dfs` isn't a self-explanatory function name